File: cpp project/source/camera.cpp
#include "stdafx.h"

Camera::Camera()
{
	 // nastaveni vsech dat
	width = 640;
	height = 480;
	fov_y = 45;
	aspect = 1.333; // 1, 1.333, 
	// matice
	eye_ = Vector3(static_cast<REAL>(-0.067), static_cast<REAL>(0.237), static_cast<REAL>(0.184));
	// matice
	tm[0][0] = static_cast<REAL>(-0.982);
	tm[0][1] = static_cast<REAL>(0.064);
	tm[0][2] = static_cast<REAL>(0.177);
	tm[1][0] = static_cast<REAL>(-0.189);
	tm[1][1] = static_cast<REAL>(-0.334);
	tm[1][2] = static_cast<REAL>(-0.924);
	tm[2][0] = static_cast<REAL>(0.0);
	tm[2][1] = static_cast<REAL>(0.940);
	tm[2][2] = static_cast<REAL>(-0.340);
}

Ray Camera::GenerateRay(const float sx, const float sy)
{
	 REAL fov_x = static_cast<REAL>(fov_y) * aspect;

	 const REAL x = ((2 * sx - (float)width) / static_cast<REAL>(width)) * tan(DEG2RAD(fov_x * 0.5));
	 
	 const REAL y = ((2 * sy - (float)height) / static_cast<REAL>(height)) * tan(DEG2RAD(fov_y * 0.5));

	 Vector3 direction = Vector3(x, -y, 1); // smer noveho paprsku
	 direction.Normalize();

	 Vector3 direction_transformed = Vector3(
		 tm[0][0]*direction.x + tm[0][1]*direction.y + tm[0][2]*direction.z,
		 tm[1][0]*direction.x + tm[1][1]*direction.y + tm[1][2]*direction.z,
		 tm[2][0]*direction.x + tm[2][1]*direction.y + tm[2][2]*direction.z);

	 // paprsek
	 Ray ray = Ray(eye_, direction_transformed);

	 // paprsek pozadi
	 Vector3 bdirection = Vector3(x, y, 1); // smer noveho paprsku
	 bdirection.Normalize();

	 Vector3 newBdirection = Vector3(
		 -1*bdirection.x + 0*bdirection.y + 0*bdirection.z,
		 0*bdirection.x + -1*bdirection.y + 0*bdirection.z,
		 0*bdirection.x + 0*bdirection.y + -0.5*bdirection.z);

	 ray.bdirection = newBdirection;

	 // vracime
	 return (ray);
}

File: static/js/factory/camera.js
raytracer.factory("Camera", [
	"Vector3",
	"Funcs",
	"Ray",
function(
	Vector3,
	Funcs,
	Ray
) {
	
	var Camera = function() {
		// sirka vyska obrazku
		this._width = 160; // 640
		this._height = 120; // 480
		// pomer
		this._aspect = 1.333;
		// uhel otoceni oka
		this._fovY = 45;
		// oko
		this._eye = new Vector3(-0.067, 0.237, 0.184);

		// transformacni matice 3x3
		this._tm = new Array(3);

		for (var i = 0; i < 3; i++) {
			this._tm[i] = new Array(3);
		}

		this._tm[0][0] = -0.982;
		this._tm[0][1] = 0.064;
		this._tm[0][2] = 0.177;
		this._tm[1][0] = -0.189;
		this._tm[1][1] = -0.334;
		this._tm[1][2] = -0.924;
		this._tm[2][0] = 0.0;
		this._tm[2][1] = 0.940;
		this._tm[2][2] = -0.340;
	};

	Camera.prototype.generateRay = function(sx, sy) {
		var fovX = this._fovY * this._aspect;

		var x = ((2 * sx - this._width) / this._width) * Math.tan(Funcs.deg2rad(fovX * 0.5));
		 
		var y = ((2 * sy - this._height) / this._height) * Math.tan(Funcs.deg2rad(this._fovY * 0.5));

		// smer noveho paprsku
		var direction = new Vector3(x, -y, 1);
		direction.normalize();

		var d = direction.rwData();

		var direction_transformed = new Vector3(
			this._tm[0][0] * d.x + this._tm[0][1] * d.y + this._tm[0][2] * d.z,
			this._tm[1][0] * d.x + this._tm[1][1] * d.y + this._tm[1][2] * d.z,
			this._tm[2][0] * d.x + this._tm[2][1] * d.y + this._tm[2][2] * d.z
		);

		// paprsek
		var ray = new Ray(this._eye, direction_transformed);

		// paprsek pozadi
		// smer noveho paprsku
		var bdirection = new Vector3(x, y, 1); 
		bdirection.normalize();

		var b = bdirection.rwData();

		var newBdirection = new Vector3(
			-1 * b.x + 0 * b.y + 0 * b.z,
			0 * b.x + -1 * b.y + 0 * b.z,
			0 * b.x + 0 * b.y + -0.5 * b.z
		);

		ray.bdirection = newBdirection;

		// vracime
		return (ray);
	};

	Camera.prototype.getWidth = function() {
		return this._width;
	};

	Camera.prototype.getHeight = function() {
		return this._height;
	};

	Camera.prototype.setWidth = function(width) {
		this._width = width;
	};

	Camera.prototype.setHeight = function(height) {
		this._height = height;
	};

	return Camera;
}]);
